File: mobile_client/lib/services/join_game_service.dart
import 'package:mobile_client/constants/enum/room_access_status.dart';

import 'room_service.dart';

class JoinGameService {
  final RoomService roomService;

  JoinGameService(this.roomService);

  Future<String?> roomIdValidator(String roomId) async {
    try {
      final status = await roomService.canPlayerJoinRoom(roomId);
      return switch (status) {
        RoomAccessStatus.opened => null,
        RoomAccessStatus.locked => "Cette salle est fermée",
        RoomAccessStatus.hidden ||
        RoomAccessStatus.friendOnly =>
          "Cette salle est cachée",
        RoomAccessStatus.deleted => "Cette salle n'existe pas",
        RoomAccessStatus.notEnoughMoney =>
          "Vous n'avez pas assez d'argent pour joindre la salle.",
        RoomAccessStatus.maxPlayerCountReached => "La salle est pleine.",
        _ => null
      };
    } catch (_) {
      return "Erreur inattendue";
    }
  }
}

File: desktop-client/src/app/home/services/join-game/join-game.service.ts
import { Injectable } from '@angular/core';
import { AbstractControl, ValidationErrors } from '@angular/forms';
import { RoomService } from '@app/home/services/room/room.service';
import { RoomAccessStatus } from '@common/enum/room-access-status';

@Injectable({
    providedIn: 'root',
})
export class JoinGameService {
    constructor(private readonly roomService: RoomService) {}

    async roomIdValidator(control: AbstractControl<string, string>): Promise<ValidationErrors | null> {
        try {
            const roomAccessStatus = await this.roomService.canPlayerJoinRoom(control.value);
            switch (roomAccessStatus) {
                case RoomAccessStatus.OPENED:
                    return null;
                case RoomAccessStatus.LOCKED:
                    return { roomIsLocked: true };
                case RoomAccessStatus.HIDDEN:
                case RoomAccessStatus.FRIEND_ONLY:
                    return { roomAccessLocked: true };
                case RoomAccessStatus.DELETED:
                    return { roomDoesNotExist: true };
                case RoomAccessStatus.NOT_ENOUGH_MONEY:
                    return { notEnoughMoney: true };
                case RoomAccessStatus.MAX_PLAYER_COUNT_REACHED:
                    return { maxPlayerCountReached: true };
            }
        } catch {
            return { otherError: true };
        }
    }
}
